coderd/prebuilds/state.go

```diff
@@ -0,0 +1,200 @@
+package prebuilds
+
+import (
+	"slices"
+	"time"
+
+	"github.com/coder/quartz"
+	"github.com/google/uuid"
+
+	"github.com/coder/coder/v2/coderd/database"
+)
+
+// CalculateState computes the current state of prebuilds for a preset, including:
+// - Actual: Number of currently running prebuilds
+// - Desired: Number of prebuilds desired as defined in the preset
+// - Eligible: Number of prebuilds that are ready to be claimed
+// - Extraneous: Number of extra running prebuilds beyond the desired count
+// - Starting/Stopping/Deleting: Counts of prebuilds in various transition states
+//
+// The function takes into account whether the preset is active (using the active template version)
+// and calculates appropriate counts based on the current state of running prebuilds and
+// in-progress transitions. This state information is used to determine what reconciliation
+// actions are needed to reach the desired state.
+func (p PresetSnapshot) CalculateState() *ReconciliationState {
+	var (
+		actual     int32
+		desired    int32
+		eligible   int32
+		extraneous int32
+	)
+
+	if p.isActive() {
+		actual = int32(len(p.Running))
+		desired = p.Preset.DesiredInstances.Int32
+		eligible = p.countEligible()
+		extraneous = max(actual-desired, 0)
+	}
+
+	starting, stopping, deleting := p.countInProgress()
+
+	return &ReconciliationState{
+		Actual:     actual,
+		Desired:    desired,
+		Eligible:   eligible,
+		Extraneous: extraneous,
+
+		Starting: starting,
+		Stopping: stopping,
+		Deleting: deleting,
+	}
+}
+
+// CalculateActions determines what actions are needed to reconcile the current state with the desired state.
+// The function:
+// 1. First checks if a backoff period is needed (if previous builds failed)
+// 2. If the preset is inactive (template version is not active), it will delete all running prebuilds
+// 3. For active presets, it calculates the number of prebuilds to create or delete based on:
+//   - The desired number of instances
+//   - Currently running prebuilds
+//   - Prebuilds in transition states (starting/stopping/deleting)
+//   - Any extraneous prebuilds that need to be removed
+//
+// The function returns a ReconciliationActions struct that will have exactly one action type set:
+// - ActionTypeBackoff: Only BackoffUntil is set, indicating when to retry
+// - ActionTypeCreate: Only Create is set, indicating how many prebuilds to create
+// - ActionTypeDelete: Only DeleteIDs is set, containing IDs of prebuilds to delete
+func (p PresetSnapshot) CalculateActions(clock quartz.Clock, backoffInterval time.Duration) (*ReconciliationActions, error) {
+	// TODO: align workspace states with how we represent them on the FE and the CLI
+	//	     right now there's some slight differences which can lead to additional prebuilds being created
+
+	// TODO: add mechanism to prevent prebuilds being reconciled from being claimable by users; i.e. if a prebuild is
```

Comment:
Don't forget to fix this and other TODOs

Reply:
We should probably implement this lock to prevent to-be-deleted prebuilds from being claimable.
Could be done in a follow-up PR.

Reply:
> prevent to-be-deleted prebuilds from being claimable 

Do we need a lock for that? the claim query already has "for update skip locked". "prebuilds being reconciled" are "in progress". We already filter for that as far as I know. I'm pretty sure we can get rid of this todo. We could write a test for this scenario if we don't already have one.

Reply:
I think Danny means that such scenario is possible:
- `CalculateActions` calculates that `prebuild 1` should be deleted, and saves this in memory in `DeleteIDs`.
- User claims `prebuild 1` because it's still claimable.
- A bit later we delete this prebuild, because it's already saved in `DeleteIDs` and basically scheduled for deletion.

<hr/>

But I guess it's rare scenario because there are only 2 cases when we delete prebuilds:
- template version is outdated or deprecated - but it means it's `not claimable`
- we have extraneous prebuilds - but I think this scenario is almost impossible?

@dannykopping so I think it's not really an issue?